base/strings/basic.jl

```diff
@@ -508,17 +508,18 @@ false
 """
 isascii(c::Char) = bswap(reinterpret(UInt32, c)) < 0x80
 isascii(s::AbstractString) = all(isascii, s)
+isascii(c::AbstractChar) = UInt32(c) < 0x80
 
 ## string map, filter, has ##
 
 function map(f, s::AbstractString)
     out = IOBuffer(sizehint=sizeof(s))
     for c in s
         c′ = f(c)
-        isa(c′, Char) || throw(ArgumentError(
-            "map(f, s::AbstractString) requires f to return Char; " *
+        isa(c′, AbstractChar) || throw(ArgumentError(
+            "map(f, s::AbstractString) requires f to return AbstractChar; " *
             "try map(f, collect(s)) or a comprehension instead"))
-        write(out, c′::Char)
+        write(out, c′::AbstractChar)
```

Comment:
Do we require all `AbstractChar` implementations to implement `write` so that it uses UTF-8? This code relies on this assumption.

Reply:
There is a fallback `write` method that works via conversion to `Char`.

Reply:
OK. But I guess it should be mentioned in the docstring for `AbstractChar` too? It could sound natural to somebody implementing e.g. a `Latin1Char` to write it in ISO-8859-1, which wouldn't be correct.

Reply:
Will do.

Reply:
Alternatively, we could define `print` to always output UTF-8, and `write` to output a raw encoded value.

Reply:
But I tend to think that the I/O encoding should be a property of the stream, not the string or character type.

Reply:
100% agree: letting the value that's being printed determine the encoding would be a mess. We do need external support for encoded streams that handle this appropriately, but `print(io, '∀')` where `io` is a UTF-16 encoded stream it should definitely not write UTF-8 data to `io` because `'∀'` is a UTF-8 encode character type. If you have a `CharU16` value and a UTF-16 encoded stream it should know that it can output code units directly, of course, but that's just a matter of providing the right `print` specializations.